### PR TITLE
[Performance fix] Make GetBlob assuming elements are cached

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Performance optimization

### PR changes
Others

### Describe
GetBlob is very often used function and majority of it is executed under critical section so it is good to optimize it as much as possible. In this PR we add assumption that objects in cache are there e.g. we instruct compiler to generate code layout in favour of situation that objects are cached. Performance improvement is **x < 1%** on mobilenetv1_int8 CLX

### CODE REVIEW

**Comment by jczaja** [comment]
@pawelpiotrowicz , @tsocha , @Silv3S, @piotrekobiIntel, @zuzg    Please review

File: paddle/fluid/platform/device_context.cc [hunk 1]
```diff
@@ -849,6 +849,15 @@ unsigned int MKLDNNDeviceContext::GetCachedObjectsNumber(void) const {
   return num_entries;
 }
 
+// TODO(jczaja): Replace with C++20 equivalents when applicable
+#ifdef _WIN32
+#define likely(expr) (expr)
+#define unlikely(expr) (expr)
+#else
+#define likely(expr) (__builtin_expect(!!(expr), 1))
```
Comment (pawelpiotrowicz) [inline paddle/fluid/platform/device_context.cc:857]: Why are using double !! ??

**Review by pawelpiotrowicz — APPROVED** [review]
LGTM

File: paddle/fluid/platform/device_context.cc (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/platform/device_context.cc:857]: __builtin_expect does accept long integer arguments. Comparison among boolean and int may raise a warning. So it was "double negated" so True is converted to 1, False to 0 and compared with long in of 1 or 0.

**Review by Silv3S — APPROVED** [review]
LGTM

**Review by jakpiase — APPROVED** [review]
LGTM

**Comment by jczaja** [comment]
@Aganlengzi Please review and merge if correct

**Review by arlesniak — APPROVED** [review]
LGTM

**Review by sfraczek — APPROVED** [review]
LGTM

**Review by ghost — APPROVED** [review]

**Review by Aganlengzi — APPROVED** [review]
LGTM

**Pull request merged**